### PR TITLE
feat: support hash value of public inputs

### PR DESCRIPTION
**Pull request by johntaiko**

### CODE REVIEW

File: zkevm-circuits/src/pi_circuit2.rs [hunk 1]
```diff
@@ -0,0 +1,1582 @@
...
+    tx_sign_hash: [u8; 32],
+}
+
+/// Extra values (not contained in block or tx tables)
+#[derive(Default, Debug, Clone)]
+pub struct ExtraValues {
+    // block_hash: H256,
+    state_root: H256,
+    prev_state_root: H256,
+}
```
Comment (Brechtpd) [inline zkevm-circuits/src/pi_circuit2.rs:93]: Ideally should be `pre_block_hash` and `block_hash`, unless we add some code to the smart contracts that allows us to put the state roots here. Also I guess the prover address needs to added to this?

File: zkevm-circuits/src/pi_circuit2.rs [hunk 2]
```diff
@@ -0,0 +1,1582 @@
...
+
+/// PublicData contains all the values that the PiCircuit recieves as input
+#[derive(Debug, Clone, Default)]
+pub struct PublicData {
+    /// chain id
+    pub chain_id: Word,
+    /// History hashes contains the most recent 256 block hashes in history,
+    /// where the latest one is at history_hashes[history_hashes.len() - 1].
+    pub history_hashes: Vec<Word>,
```
Comment (Brechtpd) [inline zkevm-circuits/src/pi_circuit2.rs:102]: Could be removed. But if it's easier to keep it let's keep it.

**Review by Brechtpd — COMMENTED** [review]
Looks good to me! I only looked at it from a high level because it would take a long time to verify if all the constraints are valid (which is not necessarily super important right now). If this does what the notion doc says then all good for me. :)

File: zkevm-circuits/src/lib.rs [hunk 3]
```diff
@@ -17,11 +17,14 @@
...
 
 pub mod bytecode_circuit;
+mod compress_circuit;
 pub mod copy_circuit;
 pub mod evm_circuit;
 pub mod exp_circuit;
 pub mod keccak_circuit;
 pub mod pi_circuit;
+pub mod pi_circuit2;
+mod rlp_circuit;
```
[the remark this answers is not in the document]
Comment (johntaiko) [reply zkevm-circuits/src/lib.rs:27]: yes, compress_circuit doesn't be used now, rlp_circuit has been combined into pi_circuit2

File: zkevm-circuits/src/table.rs [hunk 4]
```diff
@@ -1310,3 +1310,128 @@ impl<F: Field> LookupTable<F> for ExpTable {
...
+    /// True when the row is enabled
+    pub is_enabled: Column<Advice>,
+    /// Byte array input as `RLC(reversed(input))`
+    pub input_rlc: Column<Advice>, // RLC of input bytes
+    /// Byte array input length
+    pub input_len: Column<Advice>,
+    /// hash high 16 bytes
+    pub output_hi: Column<Advice>,
+    /// hash low 16 bytes
+    pub output_lo: Column<Advice>,
```
Comment (Brechtpd) [inline zkevm-circuits/src/table.rs:1326]: Nice, seems we're going for future proof here. :)
But may also make it a bit more challenging in the future to add additional circuits, but I guess that's okay. We're not trying to create a sound circuit implementation anyway so we have some flexibility.

**Pull request merged**